BSP/STM32F413/Src/BSP_CAN.c

```diff
@@ -0,0 +1,374 @@
+        GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+        GPIO_InitStruct.Alternate = GPIO_AF11_CAN3;
+        HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
+    }
+}
+
+void HAL_CAN_MspDeInit(CAN_HandleTypeDef *canHandle)
+{
+    if (canHandle->Instance == CAN1)
+    {
+        /* CAN1 clock disable */
+        __HAL_RCC_CAN1_CLK_DISABLE();
+
+        /**CAN1 GPIO Configuration
+        PA11     ------> CAN1_RX
+        PA12     ------> CAN1_TX
+        */
+        HAL_GPIO_DeInit(GPIOA, GPIO_PIN_11 | GPIO_PIN_12);
+    }
+    else if (canHandle->Instance == CAN3)
+    {
+        /* CAN3 clock disable */
+        __HAL_RCC_CAN3_CLK_DISABLE();
+
+        /**CAN3 GPIO Configuration
+        PA8     ------> CAN3_RX
+        PA15     ------> CAN3_TX
+        */
+        HAL_GPIO_DeInit(GPIOA, GPIO_PIN_8 | GPIO_PIN_15);
+    }
+}
+
+bool CAN_Send(const CAN_HandleTypeDef *can, CAN_TxPayload_t payload, bool blocking)
+{
+    /* return status */
+    int ret = true;
+
+    /* TODO: error check payload */
+
+    if (blocking) {
+        /* add payload to queue and wait till a success */
+        while (!xQueueSend(LocalCAN_TxQueue, &payload, 0)) {}
+        ret = false;
+    }
+    else {
+        /* add payload to queue and return status */
+        if (!xQueueSend(LocalCAN_TxQueue, &payload, 0))
+        {
+            ret = false;
+        }
+    }
+
+    /* TODO: if can tx is inactive, put payload into mailbox */
+
+    return ret;
+}
+
+bool CAN_SetListener(const CAN_HandleTypeDef *can, CANID_t id, QueueHandle_t *queue)
+{
+    /* set listeners depending on bus */
+    CAN_Listener_t *listeners = (can->Instance == CAN1) ? CarCAN_Listeners : LocalCAN_Listeners;
+
+    /* find if listener is already set (linear search) */
+    for (int i = 0; i < NUM_LISTENERS; i++)
+    {
+        if (listeners[i].id == id)
+        {
+            /* replace if found (don't want duplicate ids) */
+            listeners[i].queue = queue;
+            return false;
+        }
+    }
+
+    /* find empty listener spot (linear search) */
+    for (int i = 0; i < NUM_LISTENERS; i++)
+    {
+        if (listeners[i].queue == NULL)
+        {
+            /* set if found */
+            listeners[i] = (CAN_Listener_t){.id = id, .queue = queue};
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void CAN_ClearListener(const CAN_HandleTypeDef *can, CANID_t id)
+{
+    /* set listeners depending on bus */
+    CAN_Listener_t *listeners = (can->Instance == CAN1) ? CarCAN_Listeners : LocalCAN_Listeners;
+
+    /* clear listener if present */
+    for (int i = 0; i < NUM_LISTENERS; i++)
+    {
+        if (listeners[i].id == id)
+        {
+            listeners[i] = (CAN_Listener_t){0};
+        }
+    }
+}
+
+void CAN1_RX0_IRQHandler(void)
+{
+    /* disable interrupts */
+    portENTER_CRITICAL();
+
+    /* TODO: acknowledge flag? */
+
+    /* check for any pending messages */
+    while (HAL_CAN_GetRxFifoFillLevel(&CarCAN_, CAN_RX_FIFO0))
```

Comment:
is there an ISR equivalent of this function? at least for FreeRTOS we shouldn't need to do portENTER_CRITICAL directly i dont think